src/BloomBrowserUI/bookEdit/bloomField/bloomField.ts

```diff
@@ -0,0 +1,254 @@
+/// <reference path="../../lib/jquery.d.ts" />
+
+// We have a lot of trouble with the embedded Firefox's ContentEditable behavior. The chief problem is that FF *really* wants to just use <br>s,
+// which are worthless because you can't style them to, for example, do paragraph indents.
+// So this module is for various work-arounds. Eventually we should probably admit defeat and move to some existing open-source div editor.
+
+//Firefox adds a <BR> when you press return, which is lame because you can't put css styles on BR, such as indent.
+//Eventually we may use a wysiwyg add-on which does this conversion as you type, but for now, we change it when
+//you tab or click out.
+
+enum CursorPosition {start, end}
+
+class BloomField {
+
+    static ManageField(target:HTMLElement) {
+
+        BloomField.PreventRemovalOfSomeElements(target);
+
+        if(BloomField.RequiresParagraphs(target)) {
+            BloomField.ModifyForParagraphMode(target);
+            BloomField.ManageWhatHappensIfTheyDeleteEverything(target);
+            BloomField.PreventArrowingOutIntoField(target);
+            $(target).blur(function () {
+                BloomField.ModifyForParagraphMode(this);
+            });
+            $(target).focusin(function () {
+                BloomField.HandleFieldFocus(this);
+            });
+        }
+        else{
+            BloomField.PrepareNonParagraphField(target);
+        }
+    }
+
+    // Without this, ctrl+a followed by a left-arrow or right-arrow gets you out of all paragraphs,
+    // so you can start messing things up.
+    static PreventArrowingOutIntoField(field:HTMLElement) {
+        $(field).keydown(function (e) {
+            var leftArrowPressed = e.which === 37;
+            var rightArrowPressed = e.which === 39;
+            if (leftArrowPressed || rightArrowPressed) {
+                var sel = window.getSelection();
+                if (sel.anchorNode === this) {
+                    //enhance: the correct behavior would actually be to collapse the selection to one point at the beginning
+                    e.preventDefault();
+                    BloomField.MoveCursorToFirstParagraph(this, leftArrowPressed ? CursorPosition.start : CursorPosition.end);
+                }
+            }
+        })
+    }
+
+    static EnsureStartsWithParagraphElement(field:HTMLElement) {
+        if ($(field).children().length > 0 && ( $(field).children().first().prop("tagName").toLowerCase() === 'p')) {
+            return;
+        }
+        $(field).prepend('<p></p>');
+    }
+
+    static EnsureEndsWithParagraphElement(field:HTMLElement) {
+        //Enhance: move any errant paragraphs to after the imageContainer
+        if($(field).children().length > 0 && ( $(field).children().last().prop("tagName").toLowerCase() === 'p')) {
+            return;
+        }
+        $(field).append('<p></p>');
+    }
+
+    static ConvertTextNodesToParagraphs(field: HTMLElement){
```

Comment:
This is only walking the top-level nodes, so it's not going to see anything inside of paragraphs. But it sure doesn't handle <span>s and such that were created in some previous, un-paragraph-ied version. And ideally, it should. Hmm....